modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -615,14 +503,22 @@ private boolean rowIsOutOfRange(
             List<InlineIndexKeyType> keyTypes = tree.rowHandler().inlineIndexKeyTypes();
 
             for (int keyIdx = 0; keyIdx < criteriaKeysCnt; keyIdx++) {
-                RangeIndexQueryCriterion c = qry.criteria[keyIdx];
+                RangeIndexQueryCriterion c = qry.keyCond[keyIdx].range();
+                Set<IndexKey> inVals = qry.keyCond[keyIdx].inVals();
 
                 InlineIndexKeyType keyType = keyIdx < keyTypes.size() ? keyTypes.get(keyIdx) : null;
 
                 boolean descOrder = descOrderCache[keyIdx];
 
                 int maxSize = tree.inlineSize() - fieldOff;
 
+                if (inVals != null) {
+                    IndexRow row = io.getLookupRow(tree, pageAddr, idx);
```

Comment:
What do you think about using inline payload data if it is available for check IN condition?